src/functions/indAffineDirect.jl

```diff
@@ -0,0 +1,105 @@
+### CONCRETE TYPE: DIRECT PROX EVALUATION
+# prox! is computed using a QR factorization of A'.
+
+struct IndAffineDirect{R <: Real, T <: RealOrComplex{R}, M <: AbstractMatrix{T}, V <: AbstractVector{T}, F <: Factorization} <: IndAffine
```

Comment:
I think that you are right that it will fail to create a type. There should be ways to get around this by adding more type parameters.
It is easy to see that we need `res::promote_type(typeof(A), typeof(b))`, the type of the `b` we store is a bit trickier since we use norm of A to scale, with for example `b::Array{Float32,1}`, `A::Array{Complex{Float64},2}` we have to store `b::Array{Float64,1}`. This means that we need one type for `A`, one for `b`, and one for `res`, and I don't see that we get any significant improvement compared to doing something like
```julia
function IndAffineDirect{T<:RealOrComplex{<:Real}}(A::AbstratMatrix{T} b::AbstractVector)
  b2 = similar(b, T)
  b2 .= b
  IndAffineDirect(A, b2)
end
```
but then we run into problems when `b` is of "higher accuracy" than `A`. I'm not sure if there is an elegant solution without having one type for each value. But I think that if we have too many type parameters somewhere we risk that julia stops specializing on them.